Add new criterion - year

tobylane opened the pull request:
**Describe what the proposed change does**
- Add new endcondition criterion for the year. Is it clearer as complete years, without a decimal point for progress in the current year?  
The year needs 0.99 removed to work for the comparisons in checking win and lose. As with the internal values for year, it starts at 0 not 2000 and is only changed to this millennium for display.  
A requirement for over a certain year can be combined with others as a grace period, eg cure percentage must be over 60% after ten years
Examples for ending after 3 years.
#win_criteria[0].Criteria.MaxMin.Value.Group.Bound 9 1 3 1 0
#lose_criteria[0].Criteria.MaxMin.Value.Group.Bound 9 1 3 1 0

tobylane committed:
Add new criteria - year

TheCycoONE commented:
Are you using - 0.99 to get something like the floor? I don't quite understand though you described it.

Matroftt commented:
> What motivation for this feature in overall? Which benefits?

This is great, it will be possible to create, for example, such scenario which requires to cure 500 people in 3 years

tobylane commented:
> Are you using - 0.99 to get something like the floor? I don't quite understand though you described it.

The check for lose goal needs the measure to be over the target, where the win goal can just equal it. The start of year 4 is measured as 3.01, exceeding 3, causing a game to end. The wording of the tooltips is that the start of the year is needed for a win, but the end of the year for a loss. If this isn't acceptable, I'll look for an alternative solution to the measure function in get_custom_criteria.

TheCycoONE commented:
> > Are you using - 0.99 to get something like the floor? I don't quite understand though you described it.
> 
> The check for lose goal needs the measure to be over the target, where the win goal can just equal it. The start of year 4 is measured as 3.01, exceeding 3, causing a game to end. The wording of the tooltips is that the start of the year is needed for a win, but the end of the year for a loss. If this isn't acceptable, I'll look for an alternative solution to the measure function in get_custom_criteria.

It's acceptable but please comment as it's not obvious.

Alberth289346 commented:
> The year needs 0.99 removed to work for the comparisons in checking win and lose.

Why not hide that in the implementation?

If a user wants 3 years, they write "3". That's the intuitive solution to a user. If the code cannot handle that, change code rather than forcing the user to enter non-intuitive values.

Not all users read everything or understand everything. Don't assume they will do the right thing when you force them in (for them) weird solutions.  
Trying that anyway just causes an endless amount of problems with levels not working as designed.

Alberth289346 commented:
Why not use "month" or even "day"? It gives more options to the designer without much additional effort for us.

lewri commented:
> Why not use "month" or even "day"? It gives more options to the designer without much additional effort for us.

The original actually works in months for other settings. Month might be a more expected fit.

tobylane commented:
> Why not hide that in the implementation?

It is hidden, but as it needs a further explanation, it should be done another way. Eg the measure function is told it's measuring for a lose comparison, so it can adjust differently to measuring for win.

> Why not use "month" or even "day"? It gives more options to the designer without much additional effort for us.

The level writer's target needs to be a number (unless we assume they wrote something that can be treated as a date type), so it would be progress through the year. This can be converted to years and months and days in the tooltip.

lewri commented:
NB: assists with #2488

tobylane commented:
Added Date:getProgress to measure the days of the year accurately, and a tiny amount is added to this in the measure function to cause a lose end game in the correct year.  
The column height now changes each day, but the tooltips are set at dialog creation to just the year. As the dialog is closed at the end of the year for the hall of fame, it's never out of date. Rebased.

Alberth289346 commented on CorsixTH/Lua/date.lua:
```diff
@@ -48,6 +48,22 @@ local month_length = {
+  273,
+  304,
+  334,
+}
```
Don't add magic numbers that need to be kept in sync with other magic numbers.

Compute this table from the table above.

Alberth289346 commented on CorsixTH/Lua/date.lua:
```diff
@@ -263,6 +279,13 @@ function Date:isSameDay(other)
+!return (number)
+]]
+function Date:getProgress()
+  return self._year - 1 + (cumulative_days[self._month] + self._day) / 365
```
Don't use a magic `365` that depends on previous magic numbers.
Compute it.

Alberth289346 commented on CorsixTH/Lua/date.lua:
```diff
@@ -263,6 +279,13 @@ function Date:isSameDay(other)
   return self._year == other._year and self._month == other._month and self._day == other._day
 end
 
+--[[ Calculates the progress through the year as a decimal (including current day)
+!return (number)
```
What does "decimal" mean?  To me it means you write a number in base 10.

Base 10 or eg base 16 is not about the value that you return, it's just about notation when you write the value on a piece of paper.

I think what you are looking for is "as a real" or "as floating point".

What is missing here is the unit. What does `1.0` mean? If I happen to pick 10 days as unit, it matches your description, but the code doesn't do that.

Also "through the year" can be interpreted as the fraction within the current year (ie it starts at 0 again each year) rather than adding all years from the start of the game (ie cumulative).

I'd suggest to keep the explanation short, and move the more technical details about meaning of a returned value after the "!return" line.
That also nicely fixes the problem of an empty return description.

Alberth289346 commented on CorsixTH/Lua/endconditions.lua:
```diff
@@ -32,13 +32,15 @@ local local_criteria_variable = {
     -- New criteria
   {name = "staff_happiness",  icon = 16, icon_file = "MPointer", formats = 2, two_tooltips = true},
   {name = "patient_happiness",icon = 18, icon_file = "MPointer", formats = 2, two_tooltips = true},
+  {name = "year",             icon = 58, icon_file = "Font04V",  formats = 4, two_tooltips = true}, -- Starts from 0 not 2000
```
If you rename to "years_played" (or something like that), you avoid the ambiguity between 0 and 2000.

Alberth289346 replied:
Another option here is to add the base year in `Date`, and do everything in the absolute year. That may need to subtract the base year again in some computations, not sure.

Alberth289346 commented on CorsixTH/Lua/endconditions.lua:
```diff
@@ -32,13 +32,15 @@ local local_criteria_variable = {
 local get_custom_criteria = {
   staff_happiness = function(hospital) return 100 * hospital:getAverageStaffAttribute("happiness", 0.5) end,
   patient_happiness = function(hospital) return 100 * hospital:getAveragePatientAttribute("happiness", 0.5) end,
+  year = function(hospital) return hospital.world.game_date:getProgress() + 0.001 end,
```
The magic `0.001` needs a comment at least.
A new programmer is not going to understand why it's there.

I am not sure either :p

tobylane replied:
If the player has killed 10 patients on a level where a lose condition is killing 10 patients, they haven't _exceeded_ the condition yet. I think reaching day 1 of the year should count as exceeding it, but this is inconsistent with other criteria (the alternative reasoning is to say that once the player is past the first day of the year, any check of end game will result in losing, such as at the end of the year). A winning condition of kill 10 patients will be met when 10 have been killed, it only needs to be equalled, not exceeded. Adding a small number, less than a day, means that the whole number is not reached any sooner, the column height isn't changed. I'll add this as a general note to the wiki page for new criteria.

ARGAMX replied:
> If the player has killed 10 patients on a level where a lose condition is killing 10 patients, they haven't _exceeded_ the condition yet.

Another option: 
Why we we don't want shift this logic to place where comparison is happening? 
I mean in that place we can use `>`, instead of `=>` to satisfy the exceeded logic.

Alberth289346 replied:
There is a smart trick in the comparison code to check for winning and losing conditions with the same operator. That may be a contributing factor to the complications here, but not sure.

Alberth289346 commented on CorsixTH/Lua/endconditions.lua:
```diff
@@ -32,13 +32,15 @@ local local_criteria_variable = {
+  {name = "year",             icon = 58, icon_file = "Font04V",  formats = 4, two_tooltips = true}, -- Starts from 0 not 2000
 }
 
 -- A table of functions for fetching criteria values that cannot be measured
 --   directly from a hospital attribute of the same name.
```
With your addition, this comment is now false.

While at it, remove the weird indentation at the 2nd line.

Alberth289346 commented on CorsixTH/Lua/languages/english.lua:
```diff
@@ -808,6 +808,7 @@ information = {
     cheat = "Hope you didn't click the Lose Level button by accident!",
     staff_happiness = "Your average staff happiness fell below %d%.",
     patient_happiness = "Your average patient happiness fell below %d%.",
+    year = "You didn't meet the level goals within %d years.",
```
Maybe `number_of_years` ?

Alberth289346 commented on CorsixTH/Lua/dialogs/fullscreen/progress_report.lua:
```diff
@@ -82,6 +83,8 @@ function UIProgressReport:UIProgressReport(ui)
       tooltip = tooltip:format(math.floor(res_value) / 1000,
           math.floor(cur_value) / 1000)
+    elseif crit_table.formats == 4 then
+      tooltip = tooltip:format(res_value + 2000, math.floor(cur_value) + 2000)
```
Magic `2000` looks like a value that should be in `Date`.

tobylane replied:
Wherever the year is shown, 1999 is added, as that's the real magic number. I'm using 2000 because I think it's easier on level editors for 5 to mean 2005 not 2004. I don't think it's right to have two magic numbers with the same meaning, so I did neither.

ARGAMX replied:
I believe that using number without any explanation makes a number magical is the.
If there is an explanation or a variable name that gives it a name, the magic is gone.
So even if you do:
`local initial_year = 2000`
and then continue to operate on `initial_year`, there is significantly magic reducing.
And if you even write a comment to clarify why exactly 2000...
At a quick overview without context, it is not immediately obvious that 2000 is a year at all.

Alberth289346 commented:
I wrote several things about the dates that you use that don't entirely agree with each other. It is probably best if you first read all comments before starting to change things.

tobylane commented:
While I am continuing with two assumptions I made, I can see the case for reversing them. The number the level editor (person, not tool) uses to represent ingame years is 2000 off, not 1999 off as other ingame representations of the date are. The number the level editor uses should be consistent in win and lose, the same number in each means the game ends in the same year. Both of these make up for previous coding choices that didn't have a downside before.

TheCycoONE commented:
> While I am continuing with two assumptions I made, I can see the case for reversing them. The number the level editor (person, not tool) uses to represent ingame years is 2000 off, not 1999 off as other ingame representations of the date are. The number the level editor uses should be consistent in win and lose, the same number in each means the game ends in the same year. Both of these make up for previous coding choices that didn't have a downside before.

Probably want to be consistent with StartMonth/EndMonth in the level files here. It would be weird not to be.

lewri reviewed (commented):
I do apologise. But I'm still questioning why we're focussed on 'year' when the original game's files, and our custom ones, work in months.

Population: months
Visual diseases: months
Staff for hire: months
Earthquakes: months
Emergencies: months
Training (original only): months

A creator is going to expect definition values to work generally in months. Why are we bunking this now? Months are still in integer format.

Though I will also grant you the inverse of this where this feature makes sense: A player is going to work in years for a progress report. It's a troublesome division.

Generally happy to proceed on this path with caution.

TheCycoONE commented:
> I do apologise. But I'm still questioning why we're focussed on 'year' when the original game's files, and our custom ones, work in months.
> 
> Population: months Visual diseases: months Staff for hire: months Earthquakes: months Emergencies: months Training (original only): months
> 
> A creator is going to expect definition values to work generally in months. Why are we bunking this now? Months are still in integer format.
> 
> Though I will also grant you the inverse of this where this feature makes sense: A player is going to work in years for a progress report. It's a troublesome division.
> 
> Generally happy to proceed on this path with caution.

The only counter argument I can think of is that victory is every 6 months.

> I do apologise. But I'm still questioning why we're focussed on 'year' when the original game's files, and our custom ones, work in months.
> 
> Population: months Visual diseases: months Staff for hire: months Earthquakes: months Emergencies: months Training (original only): months
> 
> A creator is going to expect definition values to work generally in months. Why are we bunking this now? Months are still in integer format.
> 
> Though I will also grant you the inverse of this where this feature makes sense: A player is going to work in years for a progress report. It's a troublesome division.
> 
> Generally happy to proceed on this path with caution.

I think this is the crux of it. Every 3 months is a valid option and every other setting in the file works this way.

lewri commented on CorsixTH/Lua/dialogs/fullscreen/progress_report.lua:
```diff
@@ -40,7 +40,8 @@ function UIProgressReport:UIProgressReport(ui)
       MPointer = gfx:loadSpriteTable("Data", "MPointer"),
-      Rep02V = gfx:loadSpriteTable("QData", "Rep02V", true, palette)
+      Rep02V = gfx:loadSpriteTable("QData", "Rep02V", true, palette),
+      Font04V = gfx:loadSpriteTable("QData", "Font04V"),
```
(Arguably, no game using this should exist before the change but I'm also cautious)

TheCycoONE commented:
I suspect we can deal with the display side with a special formatting function and two or three strings %d years, %d months and another for just months, maybe one for just years. Pick based on how the math works.

tobylane commented:
```lua
    elseif crit_table.formats == 4 then -- Criterion months_played
      -- The months_played string is used if the goal is 3 years or less, or in the final year
      -- Years are shown starting at 2000
      if res_value <= 36 or (res_value - cur_value) <= 12 then
        tooltip = tooltip:format(res_value, math.floor(res_value / 12) + 2000, cur_value)
      else
        tooltip = _S.tooltip.status[direction].years_played
            :format(math.floor(res_value / 12) + 2000, math.floor(cur_value / 12) + 2000)
      end
```
In a change from the previous images, this shows the year as the user expects, with better grammar. Rebased for the checks.

Alberth289346 commented:
> This uses years, unless it's a short level or the end is near when it shows months. Is a year enough warning of a mid-year end date?

37 and 40 don't have much meaning to me. Any particular reason why you don't always break it down to year+month ?

Or even say the month name instead of "3 years, 5 months" (eg `may year 3`)?

EDIT: Possibly even "may 1st" (since "may 31st" is a lot further away, and a user may think the latter)

lewri commented on CorsixTH/Lua/languages/english.lua:
```diff
@@ -1089,10 +1090,14 @@ tooltip.status = {
   over = {
     staff_happiness = "Your average staff happiness should be over %d%. Currently it's %d%",
     patient_happiness = "Your average patient happiness should be over %d%. Currently it's %d%",
+    months_played = "You need to reach the end of %s %d. Currently it's %s %d",
```
Are there any languages where it'd be reversed? (the `%s %d` component) None come to my mind but I'm no expert.

lewri replied:
Actually, can you check if we fall into the `.0` problem here?

tobylane replied:
CJK languages may do. There's a math.floor to avoid a .0.

lewri commented on CorsixTH/Lua/dialogs/fullscreen/progress_report.lua:
```diff
@@ -40,7 +40,8 @@ function UIProgressReport:UIProgressReport(ui)
       MPointer = gfx:loadSpriteTable("Data", "MPointer"),
-      Rep02V = gfx:loadSpriteTable("QData", "Rep02V", true, palette)
+      Rep02V = gfx:loadSpriteTable("QData", "Rep02V", true, palette),
+      Font04V = gfx:loadSpriteTable("QData", "Font04V"),
```
Is Font04V in the demo files? As this new criteria is needed for the true demo experience.

tobylane replied:
The demo can't open this dialog.

lewri replied:
Yep, realised now, they just get a information box with cures in
<img width="641" height="508" alt="image" src="https://github.com/user-attachments/assets/9f79b7ea-f2bc-4b69-9feb-5a17be1a46bd" />

lewri reviewed (commented):
Nice, simpler.

lewri commented on CorsixTH/Lua/endconditions.lua:
```diff
@@ -32,13 +32,15 @@ local local_criteria_variable = {
     -- New criteria
   {name = "staff_happiness",  icon = 16, icon_file = "MPointer", formats = 2, two_tooltips = true},
   {name = "patient_happiness",icon = 18, icon_file = "MPointer", formats = 2, two_tooltips = true},
+  {name = "months_played",    icon = 46, icon_file = "Font04V",  formats = 4, two_tooltips = true},
```
I'm guessing the way this is called means Unicode doesn't overwrite this.

How does this work with the Russian edition?

lewri replied:
Tag @ARGAMX for this one
